rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+front-end the `__typename` in each result item can be used to render the 
+components accordingly.  
+
+In the "text" parameter supplied to the query there will be one special 
+operator supported, the proximity operator. Several operators are available: 
+and, or, and proximity (`&`, `|`, `<->`.) The proximity operator allows 
+one to specify max distance between search terms: 3 words apart → `<3>`. 
+
+Example query using inline fragments and the proximity operator: 
+```graphql
+query {
+  search(text: "Bob<3>run") {
+    __typename
+    ... on Band { name label { id } }
+    ... on Album { title numberOfTracks }
+    ... on Musician { name bio }
+  }
+}
+```
+
+### Tools and Design
+
+Fulltext search query system implementations often involve specific systems 
+for storing and querying the text documents; however, in an effort to reduce 
+system complexity and feature implementation time I propose starting with the 
+fulltext search features built in to PostgreSQL.
+
+A FullText search field will get its own column in the Entity table just 
+like any other field; however, the data stored will be the result of the 
+lexical, morphological analysis of text documents performed on the fields 
+included in the index. The fulltext search field will be created using the 
+Postgres ts_vector function and will be indexed using a GIN index. The subgraph 
+developer will define a ranking algorithm to be used to sort query results,
+so the end-user facing API remains easy to use without any requirement to 
+understand the ranking algorithms.   
+
+Limitations of Postgres fulltext search indexes and search algorithms: 
+  - only certain languages available (expandable with plugins like [PGroonga](https://pgroonga.github.io/)), 
+    - **languages supported out of the box**: 
+        Danish, Dutch, English, Finnish, French, German, Hungarian, Italian, 
+        Norwegian, Portuguese, Romanian, Russian, Spanish, Swedish, and Turkish.
+  - select algorithms are available, and 
+  - limited to PostgreSQL storage. 
+
+Alternative technologies (not in any particular order): 
+  - Written in Rust:
+    - [Tantivy](https://github.com/tantivy-search/tantivy)
+    - [Toshi](https://github.com/toshi-search/Toshi)
+    - [Sonic](https://github.com/valeriansaliou/sonic)
+    - [MeiliSearch](https://github.com/meilisearch/MeiliSearch)
+    - [Bayard](https://github.com/bayard-search/bayard)
+
+## Compatibility
+
+This proposal does not change any existing interfaces, so no migrations 
+will be necessary for existing subgraph deployments. 
+
+## Drawbacks and Risks
+
+There is little risk in implementing this feature.  The proposed solution 
+uses native Postgres fulltext features and there is a nonzero probability 
+this choice results in slower than optimal write and read times; 
+however the tradeoff in implementation time/complexity and the existence 
+of production use case testimonials tempers my apprehension here. 
```

Comment:
Another risk is that the fulltext search ability will not be translatable to future phases of the network.